File: web/src/api/workItems.ts
// Typed data-access functions for the /api/work-items surface (gm-xgm / M1.6).
//
// These are thin wrappers over apiFetch — React Query hooks in
// src/hooks/useWorkItems.ts own caching, invalidation, and retry. Keep this
// module side-effect-free so it composes cleanly with any query client
// (including tests that mount the hooks with a fresh QueryClient per run).

import { apiFetch } from './client';
import type { AgentRef, DefinitionOfDone, WorkItem } from '@/types/core.gen';

// ListWorkItemsEnvelope is the wire shape the gm-peg list handler emits.
// The server normalises nil slices so `items` is always a JSON array,
// never null. `total` is the pre-pagination count of items (M1.3 has
// no pagination so it equals items.length, but callers MUST NOT assume
// that once filtering / pagination lands).
export interface ListWorkItemsEnvelope {
  items: WorkItem[];
  total: number;
}

// WorkItemListFilter mirrors core.WorkItemFilter on the Go side.
// All fields optional; empty values are omitted from the query string
// so the server sees the canonical "no filter" shape.
export interface WorkItemListFilter {
  state_category?: WorkItem['state_category'][];
  kind?: string[];
  label?: string[];
  status?: string[];
  assignee_id?: string;
  sprint_id?: string;
  limit?: number;
  // gm-e12.22.1: workflow-template + wisp opt-ins. Default behavior
  // hides templates and wisps from work surfaces (Plan / Backlog /
  // Sessions / Sprints). The Workflow Library opts in via
  // include_templates; the Active runs tab opts in via include_wisps.
  include_templates?: boolean;
  include_wisps?: boolean;
  // gm-g5xz.1: ISO8601 cutoff for the /recent view. Items created
  // before this timestamp are filtered out at the bd / dolt layer.
  // Accepts RFC3339 (Date.prototype.toISOString output) or YYYY-MM-DD.
  created_since?: string;
}

// buildListQuery turns the filter into a URLSearchParams. Multi-valued
// fields are emitted as repeated params (`?kind=task&kind=epic`) —
// the server accepts both repeated and CSV, but repeated is the
// canonical wire shape since it round-trips cleanly through URL
// builders without ambiguous comma handling.
function buildListQuery(filter?: WorkItemListFilter): string {
  if (!filter) return '';
  const p = new URLSearchParams();
  for (const v of filter.state_category ?? []) p.append('state_category', v);
  for (const v of filter.kind ?? []) p.append('kind', v);
  for (const v of filter.label ?? []) p.append('label', v);
  for (const v of filter.status ?? []) p.append('status', v);
  if (filter.assignee_id) p.set('assignee_id', filter.assignee_id);
  if (filter.sprint_id) p.set('sprint_id', filter.sprint_id);
  if (filter.limit != null) p.set('limit', String(filter.limit));
  if (filter.include_templates) p.set('include_templates', 'true');
  if (filter.include_wisps) p.set('include_wisps', 'true');
  if (filter.created_since) p.set('created_since', filter.created_since);
  const qs = p.toString();
  return qs ? `?${qs}` : '';
}

// listWorkItems — GET /api/work-items. The handler returns the {items,total}
// envelope (gm-peg); this helper unwraps to a bare array so callers
// can treat listWorkItems like a query. Accepts an optional filter
// (gm-e12.9.1); omit for the unfiltered list. Use
// listWorkItemsEnvelope() below when the caller also needs `total`.
export async function listWorkItems(filter?: WorkItemListFilter): Promise<WorkItem[]> {
  const env = await apiFetch<ListWorkItemsEnvelope>(`/work-items${buildListQuery(filter)}`);
  return env.items ?? [];
}

// listWorkItemsEnvelope — same fetch, but surfaces the full envelope for
// callers that need `total` (pagination counts, empty-state copy).
export async function listWorkItemsEnvelope(
  filter?: WorkItemListFilter
): Promise<ListWorkItemsEnvelope> {
  const env = await apiFetch<ListWorkItemsEnvelope>(`/work-items${buildListQuery(filter)}`);
  return { items: env.items ?? [], total: env.total ?? 0 };
}

// getWorkItem — GET /api/work-items/{id}. Returns one WorkItem with its full
// Relationship graph plus any adaptor-native edges under
// Custom["beads:dependencies"] (see internal/server/work_items.go, gm-kn2).
// Throws ApiError with status 404 / code "session_not_found" when the
// id is unknown.
export async function getWorkItem(id: string): Promise<WorkItem> {
  if (!id) {
    throw new Error('getWorkItem: id is required');
  }
  return apiFetch<WorkItem>(`/work-items/${encodeURIComponent(id)}`);
}

// WorkItemPatch mirrors the Go shape (internal/core/workplane.go).
// Every field is optional; the wire encoding is `omitempty`-friendly,
// so undefined fields stay out of the JSON body and the adaptor
// treats unset == "no change".
export interface WorkItemPatch {
  title?: string;
  description?: string;
  status?: string;
  state_category?: WorkItem['state_category'];
  priority?: number | null;
  labels?: string[];
  owner?: AgentRef | null;
  assignee?: AgentRef | null;
  sprint_id?: string | null;
  // parent_id re-parents the work item via its parent_child edge.
  // Three-state sentinel (gm-gsbj) matching the Go *string on the
  // server: undefined / omitted = no change; "" (empty string) =
  // clear (orphan); non-empty string = reparent under that id.
  // JSON null is NOT used — it decodes to a nil Go pointer, which is
  // the "no change" sentinel, not "clear".
  parent_id?: string;
  dod?: DefinitionOfDone | null;
  // custom is sent as the full map the adaptor should persist. The
  // drawer's Extensions editor (gm-root.13) writes a single key by
  // spreading the existing item.custom and overwriting one entry, so
  // adaptors that treat the field as a replacement don't lose the
  // rest of the extension map.
  custom?: Record<string, unknown>;
}

// Confirm header MUST match internal/server/nonce.go ConfirmHeader.
// Kept on the api/ side so every mutation route uses the same token
// without duplicating the literal across hooks.
export const CONFIRM_HEADER = 'X-GEMBA-Confirm';

// CreateWorkItemInput is the caller-facing shape for createWorkItem.
// Mirrors the fields the boundary decoder (transport.DecodeCreateWorkItem)
// accepts: required kind/title/status/state_category + optional
// description, priority, labels, assignee, relationships. Server-assigned
// fields (id, timestamps) are not accepted here — the decoder rejects
// them if sent.
export interface CreateWorkItemInput {
  kind: string;
  title: string;
  status: string;
  state_category: WorkItem['state_category'];
  description?: string;
  priority?: number | null;
  labels?: string[];
  assignee?: AgentRef | null;
  // Parent is encoded as a parent_child Relationship with To="" because
  // the new bead's id isn't known until the adaptor assigns it. The bd
  // adaptor translates this to `bd create --parent <id>`; adaptors that
  // set parent differently will receive the same relationship and can
  // translate as they see fit.
  relationships?: { kind: string; from: string; to: string }[];
}

// createWorkItem — POST /api/work-items (gm-e12.10). Nonce-gated like
// updateWorkItem; a caller passing the same nonce twice gets the same
// cached 201 envelope back, not a duplicate bead.
//
// Server response is the materialized WorkItem with its backend-assigned
// id and timestamps. Callers should pipe the return value into the
// react-query cache (the useCreateWorkItem hook does this).
export async function createWorkItem(
  input: CreateWorkItemInput,
  opts: { nonce?: string } = {}
): Promise<WorkItem> {
  if (!input.title) {
    throw new Error('createWorkItem: title is required');
  }
  return apiFetch<WorkItem>('/work-items', {
    method: 'POST',
    body: JSON.stringify({ item: input }),
    headers: {
      'Content-Type': 'application/json',
      [CONFIRM_HEADER]: opts.nonce ?? freshNonce(),
    },
  });
}

// updateWorkItem — PATCH /api/work-items/{id}. Generates a fresh UUID nonce
// per call so a SPA double-click can't double-apply (the server
// caches replays per nonce and returns the cached envelope verbatim).
// Caller can pass an explicit `nonce` for retry semantics — useful
// when a request is in-flight and the network drops; resending with
// the same nonce is idempotent.
export async function updateWorkItem(
  id: string,
  patch: WorkItemPatch,
  opts: { nonce?: string } = {}
): Promise<WorkItem> {
  if (!id) {
    throw new Error('updateWorkItem: id is required');
  }
  return apiFetch<WorkItem>(`/work-items/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
    headers: {
      'Content-Type': 'application/json',
      [CONFIRM_HEADER]: opts.nonce ?? freshNonce(),
    },
  });
}

// freshNonce returns a UUID-like opaque token. crypto.randomUUID is
// available in every browser the SPA targets and in jsdom; the
// fallback covers older test environments without crypto wired up.
function freshNonce(): string {
  if (typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function') {
    return crypto.randomUUID();
  }
  // Math.random fallback — only the test environment hits this.
  return `nonce-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
}

File: internal/server/work_items.go
package server

import (
	"errors"
	"net/http"
	"net/url"
	"strconv"
	"strings"

	"github.com/MikeBengtson/gemba/internal/core"
	"github.com/MikeBengtson/gemba/internal/server/httperr"
	"github.com/go-chi/chi/v5"
)

// listWorkItems is the GET /api/work-items handler. It calls the
// registered WorkPlane's ListWorkItems with a zero-valued filter (no
// narrowing — filtering and pagination are deferred to later
// milestones) and returns the envelope `{items: []WorkItem, total: N}`.
// The SPA's useWorkItems() hook drives off this shape.
//
// Error envelopes follow the shared httperr contract (gm-root.1.1):
//
//	503 {"error": "adaptor_not_configured", ...} — no host/WorkPlane wired
//	503 {"error": "adaptor_degraded", ...}       — tagged AdaptorError
//	500 {"error": "internal", ...}               — untagged adaptor error
//
// An empty-but-healthy adaptor surfaces as 200 with `{items: [], total: 0}`;
// the handler normalises nil slices so the wire shape is a JSON array, not null.
func (r *Router) listWorkItems(w http.ResponseWriter, req *http.Request) {
	if r.host == nil {
		httperr.Write(w, http.StatusServiceUnavailable,
			"adaptor_not_configured", "no WorkPlane adaptor registered")
		return
	}
	wp := r.host.WorkPlane()
	if wp == nil {
		httperr.Write(w, http.StatusServiceUnavailable,
			"adaptor_not_configured", "no WorkPlane adaptor registered")
		return
	}

	filter, err := parseWorkItemFilter(req.URL.Query())
	if err != nil {
		httperr.Write(w, http.StatusBadRequest, "bad_request", err.Error())
		return
	}

	items, err := wp.ListWorkItems(req.Context(), filter)
	if err != nil {
		httperr.WriteError(w, err)
		return
	}
	if items == nil {
		items = []core.WorkItem{}
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"items": items,
		"total": len(items),
	})
}

// parseWorkItemFilter turns the request query string into a
// core.WorkItemFilter (gm-e12.9.1). Multi-value fields accept either
// repeated params (?state_category=backlog&state_category=unstarted)
// or comma-separated lists (?state_category=backlog,unstarted) — pick
// whichever reads better at the call site.
//
// Invalid state_category / kind values return an error so typos
// surface as 400 rather than silently returning an empty list.
func parseWorkItemFilter(q url.Values) (core.WorkItemFilter, error) {
	var f core.WorkItemFilter
	f.Kinds = csvList(q["kind"])
	f.Statuses = csvList(q["status"])
	f.Labels = csvList(q["label"])
	f.IDs = toWorkItemIDs(csvList(q["id"]))

	for _, raw := range csvList(q["state_category"]) {
		sc := core.StateCategory(raw)
		if !sc.Valid() {
			return core.WorkItemFilter{}, &badRequestError{
				msg: "unknown state_category: " + raw,
			}
		}
		f.StateCategory = append(f.StateCategory, sc)
	}

	if v := strings.TrimSpace(q.Get("assignee_id")); v != "" {
		id := core.AgentID(v)
		f.AssigneeID = &id
	}
	if v := strings.TrimSpace(q.Get("sprint_id")); v != "" {
		s := v
		f.SprintID = &s
	}
	if v := strings.TrimSpace(q.Get("limit")); v != "" {
		n, err := strconv.Atoi(v)
		if err != nil || n < 0 {
			return core.WorkItemFilter{}, &badRequestError{
				msg: "limit must be a non-negative integer",
			}
		}
		f.Limit = n
	} else {
		// Default to a high cap when the caller doesn't specify (gm-nr67).
		// The bd CLI's own default is 50, which silently hides newer /
		// lower-priority items from the SPA's Board + Grid (which call
		// without a limit). 1000 covers every realistic workspace today
		// without unbounded blast radius. Proper pagination ships in a
		// later milestone; this is the interim cap.
		f.Limit = defaultListLimit
	}
	return f, nil
}

// defaultListLimit is the cap parseWorkItemFilter applies when the
// caller didn't specify ?limit=. Picked to be larger than any realistic
// workspace a single-tenant gemba instance manages today (cf. gm-e12.3
// virtualised grid is 10k rows; we sit one order of magnitude under
// that so a missing limit can't tank the server).
const defaultListLimit = 1000

// csvList flattens a slice of query values where each element may
// itself be a comma-separated list. Empty entries are dropped.
func csvList(in []string) []string {
	var out []string
	for _, raw := range in {
		for _, part := range strings.Split(raw, ",") {
			part = strings.TrimSpace(part)
			if part != "" {
				out = append(out, part)
			}
		}
	}
	return out
}

func toWorkItemIDs(in []string) []core.WorkItemID {
	if len(in) == 0 {
		return nil
	}
	out := make([]core.WorkItemID, 0, len(in))
	for _, s := range in {
		out = append(out, core.WorkItemID(s))
	}
	return out
}

// badRequestError is a typed wrapper so parseWorkItemFilter returns a
// single error shape the handler maps to 400.
type badRequestError struct{ msg string }

func (e *badRequestError) Error() string { return e.msg }

// getWorkItem is the GET /api/work-items/{id} handler. Returns a single
// WorkItem with its full Relationship graph (blocks / parent_child /
// relates_to) plus any extension edges the adaptor surfaces on
// Custom["beads:dependencies"] (or whatever the active adaptor names
// its native-edge custom key). Drives the SPA drill-in drawer.
//
// Error envelopes go through the shared httperr package so every data
// handler emits the same wire shape (gm-root.1.1):
//
//	400 {"error": "bad_request", "message": "missing work item id"}
//	404 {"error": "session_not_found", "message": "work item gm-foo not found"}
//	503 {"error": "adaptor_not_configured" | "adaptor_degraded", ...}
//	500 {"error": "internal", "message": "..."}
//
// The handler itself does not walk the relationship graph — the bound
// WorkPlane adaptor is responsible for populating Relationships,
// Evidence, and Custom fields before returning.
func (r *Router) getWorkItem(w http.ResponseWriter, req *http.Request) {
	raw := chi.URLParam(req, "id")
	if raw == "" {
		httperr.Write(w, http.StatusBadRequest, "bad_request", "missing work item id")
		return
	}
	// chi returns the path param verbatim; adaptors that prefix ids with
	// a workspace segment (the dolt adaptor's "gemba/gemba/" prefix) mean
	// the SPA sends %2F-encoded slashes. Decode here so the adaptor
	// receives the canonical id and not the still-encoded wire form.
	id, err := url.PathUnescape(raw)
	if err != nil {
		httperr.Write(w, http.StatusBadRequest, "bad_request",
			"malformed work item id: "+err.Error())
		return
	}

	if r.host == nil {
		httperr.Write(w, http.StatusServiceUnavailable,
			"adaptor_not_configured", "no WorkPlane adaptor registered")
		return
	}
	wp := r.host.WorkPlane()
	if wp == nil {
		httperr.Write(w, http.StatusServiceUnavailable,
			"adaptor_not_configured", "no WorkPlane adaptor registered")
		return
	}

	item, err := wp.GetWorkItem(req.Context(), core.WorkItemID(id))
	if err != nil {
		// Tag the bare sentinel with the work item id so the 404 envelope
		// carries a self-describing message. Adaptors that already
		// return a *core.AdaptorError flow through to httperr.WriteError
		// unchanged.
		if errors.Is(err, core.ErrNotFound) && core.AsAdaptorError(err) == nil {
			err = core.NewAdaptorError(core.KindSessionNotFound, "work item %s not found", id)
		}
		httperr.WriteError(w, err)
		return
	}

	writeJSON(w, http.StatusOK, item)
}
